File: src/main/java/com/vaadin/example/ui/FilterCheckboxes.java
package com.vaadin.example.ui;

import com.vaadin.example.search.Categories;
import com.vaadin.flow.component.checkbox.Checkbox;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;

import java.util.*;

public class FilterCheckboxes extends HorizontalLayout {
    private Map<Categories, Checkbox> checkboxes = new EnumMap<>(Categories.class);

    public FilterCheckboxes() {
        addClassName("searchPanel");
        setWidth("100%");
        for (Categories category : Categories.values()) {
            Checkbox checkbox = new Checkbox(category.getLabel());
            checkboxes.put(category, checkbox);
            add(checkbox);
        }
    }

    public List<String> getFilter() {
        List<String> result = new ArrayList<>();
        checkboxes.forEach((category, checkbox) -> {
            if (checkbox.getValue())
                result.add(category.getLabel());
        });
        return result;
    }
}

File: src/main/kotlin/com/vaadin/example/ui/FilterCheckboxes.kt
package com.vaadin.example.ui

import com.vaadin.example.search.Categories
import com.vaadin.flow.component.checkbox.Checkbox
import com.vaadin.flow.component.orderedlayout.HorizontalLayout

class FilterCheckboxes(private val searchPanel: SearchPanel) : HorizontalLayout() {
    private val checkboxes = HashMap<Categories, Checkbox>()

    init {
        setWidthFull()
    }

    fun refresh(categories : Map<Categories, Int>) {
        checkboxes.clear()
        removeAll()
        categories.forEach {
            val checkbox = Checkbox("${it.key.label} (${it.value})")
            checkbox.addValueChangeListener { searchPanel.showItems(getFilter()) }
            checkboxes[it.key] = checkbox
            add(checkbox)
        }
    }

    private fun getFilter() : Set<Categories> {
        return checkboxes.entries.filter { it.value.value }.map { it.key }.toSet()
    }
}